src/contracts/middleware/BLSRegistryCoordinatorWithIndices.sol

```diff
@@ -0,0 +1,555 @@
+                uint96 registeringOperatorStake = stakeRegistry.getCurrentOperatorStakeForQuorum(operatorIdsToSwap[0], quorumNumber);
+
+                // check the registering operator has more than the kick BIPs of the operator to kick's stake
+                require(
+                    registeringOperatorStake > operatorToKickStake * operatorSetParam.kickBIPsOfOperatorStake / BIPS_DENOMINATOR,
+                    "BLSRegistryCoordinatorWithIndices.registerOperatorWithCoordinator: registering operator has less than kickBIPsOfOperatorStake"
+                );
+                
+                // check the that the operator to kick has less than the kick BIPs of the total stake
+                require(
+                    operatorToKickStake < totalStakeForQuorum * operatorSetParam.kickBIPsOfTotalStake / BIPS_DENOMINATOR,
+                    "BLSRegistryCoordinatorWithIndices.registerOperatorWithCoordinator: operator to kick has more than kickBIPSOfTotalStake"
+                );
+            }
+            
+            // kick the operator
+            _deregisterOperatorWithCoordinator(
+                operatorKickParams[i].operator, 
+                quorumNumbers[i:i+1], 
+                operatorKickParams[i].pubkey, 
+                operatorIdsToSwap
+            );
+        }
+    }
+
+    /**
+     * @notice Deregisters the msg.sender as an operator from the middleware
+     * @param quorumNumbers are the bytes representing the quorum numbers that the operator is registered for
+     * @param deregistrationData is the the data that is decoded to get the operator's deregistration information
+     * @dev `deregistrationData` should be a tuple of the operator's BLS public key, the list of operator ids to swap
+     */
+    function deregisterOperatorWithCoordinator(
+        bytes calldata quorumNumbers,
+        bytes calldata deregistrationData
+    ) external onlyWhenNotPaused(PAUSED_DEREGISTER_OPERATOR) {
+        // get the operator's deregistration information
+        (BN254.G1Point memory pubkey, bytes32[] memory operatorIdsToSwap) 
+            = abi.decode(deregistrationData, (BN254.G1Point, bytes32[]));
+        // call internal function to deregister the operator
+        _deregisterOperatorWithCoordinator(msg.sender, quorumNumbers, pubkey, operatorIdsToSwap);
+    }
+
+    /**
+     * @notice Deregisters the msg.sender as an operator from the middleware
+     * @param quorumNumbers are the bytes representing the quorum numbers that the operator is registered for
+     * @param pubkey is the BLS public key of the operator
+     * @param operatorIdsToSwap is the list of the operator ids to swap the index of the operator with in each 
+     * quorum when removing the operator from the quorum's ordered list. The provided operator ids should be the 
+     * those of the operator's with the largest index in each quorum that the operator is deregistering from, in
+     * ascending order of quorum number.
```

Comment:
do we not keep track of the total number of operators for each quorum?
if not, makes sense. otherwise I'd assume we could look this info up.
I'll try to circle back on this.

Reply:
we store the total number of operators, but we don't store a list of operators

we store (address => index) not (index => address)

Reply:
ahhh, gotcha! OK makes sense. We know the max indices but not the corresponding operator IDs 👍